### PR TITLE
chore(ZNTA-1413): use Takari for faster builds

### PR DESCRIPTION
**Pull request by seanf**
Use Takari for incremental, parallel builds: https://zanata.atlassian.net/browse/ZNTA-1413

Use build-tools.jar to share checkstyle/findbugs config: https://zanata.atlassian.net/browse/ZNTA-1218

## Reviewer tasks

The following is based on the
[Zanata Development Guidelines](https://github.com/zanata/zanata-server/wiki/Development-Guidelines).
As a reviewer, you should check the guidelines regularly for updates, and just
to refresh your memory.

The reviewer can use this list for reference to make sure
they have considered all the important points.

Use the checkboxes or not, whatever works best for you.
- Code quality
  - [ ] Code passes style check (checkstyle/eslint)
  - [ ] Code is clear and concise
  - [ ] UI code is internationalised
  - [ ] Code has adequate comments where appropriate
  - Code is in an appropriate place
    - [ ] Classes are in appropriate packages
    - [ ] Code fits with class's responsibilities (SRP)
  - [ ] Configuration files are documented enough
  - If code is removed
    - [ ] No remaining code references the removed code
      - [ ] No orphaned rewrite rules
    - [ ] No orphaned config settings
    - [ ] No remaining comments refer to the removed code
    - [ ] Unused imports and libraries are removed
- Testing
  - [ ] New code has tests
  - [ ] Changed code has tests
  - [ ] All tests pass
  - [ ] Test coverage stable or increased
- Documentation
  - [ ] New features are documented
  - [ ] Docs removed for deleted features
  - [ ] Docs updated for all changed features
  - [ ] Developer/sysadmin docs updated if appropriate
- If there are new dependencies
  - [ ] Does each new dependency pass all the
     [Considerations for new dependencies](https://github.com/zanata/zanata-server/wiki/Development-Guidelines#new-technologies-and-dependencies)?

---

_This template can be updated in .github/PULL_REQUEST_TEMPLATE.md_

---

This change is [Reviewable](https://reviewable.io/reviews/zanata/zanata-platform/18)

### CODE REVIEW

**Comment by huangp** [comment]
Reviewed 36 of 36 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[parent/checkstyle-suppressions.xml, line 16 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwqIDWsnTj3H_SZYdx:-KVwqIDWsnTj3H_SZYdy:bz11yxk) ([raw file](https://github.com/zanata/zanata-platform/blob/9abfe4fd7190e5dce5bac96ea1fac99f1b341e26/parent/checkstyle-suppressions.xml#L16)):*
> ```XML
>   
>   <suppress checks=".*" files="db\.changelog-(1\..a?|2\..|3\.[01])\.xml$" />
>   <suppress checks=".*" files="db/.*_baseline\.sql$" />
> ```

Are these no longer a problem? We still have codemirror cache js file under src/main/resources

---

*[server/functional-test/src/test/conf/zanata-ds.xml, line 15 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwsZnGoSpc7znXT0mU:-KVwsZnGoSpc7znXT0mV:bpn202h) ([raw file](https://github.com/zanata/zanata-platform/blob/2bfcf13f340690b688f9b25c5d307cad94ed4216/server/functional-test/src/test/conf/zanata-ds.xml#L15)):*
> ```XML
>     use-ccm="true"
>     use-java-context="true">
>     <connection-url>@ds.connection.url@</connection-url>
> ```

Can you explain what is this ```@xxx@``` thing for?

---

*[server/zanata-frontend/pom.xml, line 173 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwuMaXReHOha6TnvNe:-KVwuMaXReHOha6TnvNf:b-izz176) ([raw file](https://github.com/zanata/zanata-platform/blob/9abfe4fd7190e5dce5bac96ea1fac99f1b341e26/server/zanata-frontend/pom.xml#L173)):*
> ```XML
>             </goals>
>             <configuration>
>               <skip>${skipTests}</skip>
> ```

Don't we want this?

---

*[server/zanata-frontend/pom.xml, line 245 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwukOTWpicvyoxFoJc:-KVwukOTWpicvyoxFoJd:b-2b6t3z) ([raw file](https://github.com/zanata/zanata-platform/blob/2bfcf13f340690b688f9b25c5d307cad94ed4216/server/zanata-frontend/pom.xml#L245)):*
> ```XML
>               <directory>src</directory>
>               <includes>
>                 <include>*/node_modules/**</include>
> ```

What about the stuff under src/editor/app/dist?

---

*[server/zanata-test-war/pom.xml, line 171 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwvKi95gwtQzwumaFw:-KVwvKi95gwtQzwumaFx:byj4k2k) ([raw file](https://github.com/zanata/zanata-platform/blob/2bfcf13f340690b688f9b25c5d307cad94ed4216/server/zanata-test-war/pom.xml#L171)):*
> ```XML
>       </plugin>
>       <plugin>
>         <artifactId>maven-war-plugin</artifactId>
> ```

Before this change, if we skipFuncTests then this module won't do much but looks like now it will do war overlay regardless whether we are going to run functional test. Is this true?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/18)*

**Comment by seanf** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[parent/checkstyle-suppressions.xml, line 16 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwqIDWsnTj3H_SZYdx:-KVwx1qLVYa5hn3QyUYg:b-r4slzl) ([raw file](https://github.com/zanata/zanata-platform/blob/9abfe4fd7190e5dce5bac96ea1fac99f1b341e26/parent/checkstyle-suppressions.xml#L16)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary>
> Are these no longer a problem? We still have codemirror cache js file under src/main/resources
</details>
No, they don't seem to be.

I was running into performance problems with the checkstyle plugin - when it tries to scan all the files under node_modules, it takes an incredibly long time. Suppressing the errors doesn't help, because it apparently reads all the files.

So I have had had to tweak the includes pattern. Putting node_modules into excludes didn't work. Might be a bug in the plugin.

But now I'm worried that checkstyle didn't check anything, because all the "audits" are very quick. I haven't found a way of getting it to log the filenames it checks...

I'll have to check on this.

---

*[parent/pom.xml, line 263 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwxhobHkXT2TXPULCe:-KVwxhobHkXT2TXPULCf:b7lfg1s) ([raw file](https://github.com/zanata/zanata-platform/blob/2bfcf13f340690b688f9b25c5d307cad94ed4216/parent/pom.xml#L263)):*
> ```XML
>           
>           <includes>*.xml,*.txt,*.md,etc/**,src/main/**,src/test/**,src/etc/**,src/conf/**</includes>
>           <excludes>src/test/resources/**</excludes>
> ```

@huangp note that the includes pattern has changed here. This is why we don't need so many suppressions.

---

*[server/functional-test/src/test/conf/zanata-ds.xml, line 15 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwsZnGoSpc7znXT0mU:-KVwzY7ezb8yE5EiKO7s:b59h8em) ([raw file](https://github.com/zanata/zanata-platform/blob/2bfcf13f340690b688f9b25c5d307cad94ed4216/server/functional-test/src/test/conf/zanata-ds.xml#L15)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary>
> Can you explain what is this ```@xxx@``` thing for?
</details>
It's just an alternative to `${*}` which supported by Maven for interpolation. The important thing is that it is not supported by JBoss/WildFly, which means you can distinguish between Maven property interpolation and JBoss interpolation.

We previously had `${*}` in standalone.xml where some of the values were replaced at build time by Maven, and some at runtime by JBoss. Now it's obvious which are which in that file. This file (zanata-ds.xml) was just changed to use the same approach, even though it doesn't have the same mixture.

---

*[server/zanata-frontend/pom.xml, line 173 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwuMaXReHOha6TnvNe:-KVx-tXpPpbEvBP9rJyz:b-7s3iku) ([raw file](https://github.com/zanata/zanata-platform/blob/9abfe4fd7190e5dce5bac96ea1fac99f1b341e26/server/zanata-frontend/pom.xml#L173)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary>
> Don't we want this?
</details>
We don't need it. The npm goal accepts the property skipTests directly.

---

*[server/zanata-frontend/pom.xml, line 245 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwukOTWpicvyoxFoJc:-KVx-KSShA2fB7yq1jm2:bbugkqz) ([raw file](https://github.com/zanata/zanata-platform/blob/2bfcf13f340690b688f9b25c5d307cad94ed4216/server/zanata-frontend/pom.xml#L245)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary>
> What about the stuff under src/editor/app/dist?
</details>
Good point. Thanks.

Done.

---

*[server/zanata-test-war/pom.xml, line 171 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwvKi95gwtQzwumaFw:-KVx03WaXINjy5R-TrNY:b-q8mbh4) ([raw file](https://github.com/zanata/zanata-platform/blob/2bfcf13f340690b688f9b25c5d307cad94ed4216/server/zanata-test-war/pom.xml#L171)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary>
> Before this change, if we skipFuncTests then this module won't do much but looks like now it will do war overlay regardless whether we are going to run functional test. Is this true?
</details>
Yes, but only if you let the build go past zanata-war. If you don't want functional tests, why build zanata-test-war or functional tests at all?

The old way was really error-prone, because it was depending on a build artifact in another module's target/ directory, instead of using the Maven dependency system. It was basically subverting the maven module dependency/reactor system, and not compatible with the idea of a smart builder scheduling builds in parallel. It was trying to build functional test in parallel with zanata-war, before zanata-test-war.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/18)*

**Comment by huangp** [comment]
Review status: 32 of 33 files reviewed at latest revision, 2 unresolved discussions.

---

*[parent/pom.xml, line 263 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwxhobHkXT2TXPULCe:-KVxADSOx-efdycDTI1G:bcy8plc) ([raw file](https://github.com/zanata/zanata-platform/blob/2bfcf13f340690b688f9b25c5d307cad94ed4216/parent/pom.xml#L263)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> @huangp note that the includes pattern has changed here. This is why we don't need so many suppressions.
</details>
Yes I notice that. But still src/main/resource is supposed to be included and AFAIK codemirror file is under it. I would've thought it would fail the extra newline check?

---

*[server/functional-test/src/test/conf/zanata-ds.xml, line 15 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwsZnGoSpc7znXT0mU:-KVxAn3cQlqvM8jw0bXo:b-dnw5lq) ([raw file](https://github.com/zanata/zanata-platform/blob/2bfcf13f340690b688f9b25c5d307cad94ed4216/server/functional-test/src/test/conf/zanata-ds.xml#L15)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> It's just an alternative to `${*}` which supported by Maven for interpolation. The important thing is that it is not supported by JBoss/WildFly, which means you can distinguish between Maven property interpolation and JBoss interpolation.
> 
> We previously had `${*}` in standalone.xml where some of the values were replaced at build time by Maven, and some at runtime by JBoss. Now it's obvious which are which in that file. This file (zanata-ds.xml) was just changed to use the same approach, even though it doesn't have the same mixture.
</details>
I don't understand how maven can interpolate them as they are now under src/test/conf. Will cargo do that when it copy the file over?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/18)*

**Comment by seanf** [comment]
Review status: 32 of 33 files reviewed at latest revision, 2 unresolved discussions.

---

*[parent/pom.xml, line 263 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwxhobHkXT2TXPULCe:-KVxKcz6KA0rgaBja1JV:ba95g99) ([raw file](https://github.com/zanata/zanata-platform/blob/2bfcf13f340690b688f9b25c5d307cad94ed4216/parent/pom.xml#L263)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary>
> Yes I notice that. But still src/main/resource is supposed to be included and AFAIK codemirror file is under it. I would've thought it would fail the extra newline check?
</details>
Yes, I need to look into the checkstyle business.

---

*[server/functional-test/src/test/conf/zanata-ds.xml, line 15 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwsZnGoSpc7znXT0mU:-KVxJuXmKluBNS3niKJK:b-m9kgkf) ([raw file](https://github.com/zanata/zanata-platform/blob/2bfcf13f340690b688f9b25c5d307cad94ed4216/server/functional-test/src/test/conf/zanata-ds.xml#L15)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary>
> I don't understand how maven can interpolate them as they are now under src/test/conf. Will cargo do that when it copy the file over?
</details>
Oh, sorry. I forgot I had used that. Yes, Cargo does it too. The good thing then is that we can control which properties will or won't be interpolated (instead of just grabbing all Maven's system properties). Plus the interpolation will always be up to date, because it's done when cargo is invoked.

* https://codehaus-cargo.github.io/cargo/Maven2+Plugin+Reference+Guide.html
* https://codehaus-cargo.github.io/cargo/Configuration+files+option.html

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/18)*

**Comment by seanf** [comment]
Review status: 31 of 33 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[parent/checkstyle-suppressions.xml, line 16 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KVwqIDWsnTj3H_SZYdx:-KVxlnFK3GfUzBZDSkyk:b-qvbbn9) ([raw file](https://github.com/zanata/zanata-platform/blob/9abfe4fd7190e5dce5bac96ea1fac99f1b341e26/parent/checkstyle-suppressions.xml#L16)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> No, they don't seem to be.
> 
> I was running into performance problems with the checkstyle plugin - when it tries to scan all the files under node_modules, it takes an incredibly long time. Suppressing the errors doesn't help, because it apparently reads all the files.
> 
> So I have had had to tweak the includes pattern. Putting node_modules into excludes didn't work. Might be a bug in the plugin.
> 
> But now I'm worried that checkstyle didn't check anything, because all the "audits" are very quick. I haven't found a way of getting it to log the filenames it checks...
> 
> I'll have to check on this.
</details>
I thought I was testing with my modified suppression file, but apparently not. I've put the suppressions back (except zanata-spa).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/18)*

**Comment by huangp** [comment]
:lgtm: -(Awaiting review) +Reviewed +(On QA) 

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-:-KW01vgrjbI3-JBo010U:b49ms4)*

**Comment by seanf** [comment]
@huangp Could I get you to review another commit please? It's just the checkstyle suppressions file. The rest of it is a straight rebase of the existing commits.

**Comment by huangp** [comment]
:lgtm:

---

Reviewed 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[parent/checkstyle-suppressions.xml, line 14 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KW0L-w1q6jg6sdsDvU7:-KW0L-w1q6jg6sdsDvU8:b10qe12) ([raw file](https://github.com/zanata/zanata-platform/blob/f6331a69ee857dd3f775d885210fca432189cadd/parent/checkstyle-suppressions.xml#L14)):*
> ```XML
>   
>   <suppress checks=".*" files="db\.changelog-(1\..a?|2\..|3\.[01])\.xml$" />
>   <suppress checks=".*" files="db[/\\].*_baseline\.sql$" />
> ```

Is this regex change to make it OS friendly? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-:-KW0LF9c7glSOttw0jRo:bnfp4nl)*

**Comment by seanf** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[parent/checkstyle-suppressions.xml, line 14 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-KW0L-w1q6jg6sdsDvU7:-KW0O0XAUG3o6_hD4Oos:baemjc7) ([raw file](https://github.com/zanata/zanata-platform/blob/f6331a69ee857dd3f775d885210fca432189cadd/parent/checkstyle-suppressions.xml#L14)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary>
> Is this regex change to make it OS friendly? 
</details>
Yes, Reviewable should still show you the commit messages somewhere: "Fix checkstyle suppression paths for Windows"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/18)*

**Comment by huangp** [comment]
:lgtm:

---

Reviewed 1 of 36 files at r1, 1 of 1 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-:-KW0X0yCqgDQJP2BCWmf:bnfp4nl)*

**Comment by djansen-redhat** [comment]
@zanata-jenkins retest this please

**Comment by djansen-redhat** [comment]
@seanf I'm finding that frontend fails - do you use -DexcludeFrontend when you build?

File: server/zanata-frontend/pom.xml [hunk 1]
```diff
@@ -46,46 +45,52 @@
...
+            <goals>
+              <goal>execute</goal>
+            </goals>
+          </execution>
+        </executions>
+        <!-- check for ignore-scripts = true in .npmrc, which breaks npm -->
+        <configuration>
+          <scripts>
+            <script>
+              def ignoreScripts = new File(System.getProperty('user.home'), '.npmrc').text.readLines().any{it.matches(/\s*ignore-scripts\s*=\s*true\s*/)}
```
Comment (djansen-redhat) [inline server/zanata-frontend/pom.xml]: This seems to look for a $HOME/.npmrc which does not exist

Comment (seanf) [inline server/zanata-frontend/pom.xml]: @djansen-redhat Yep, I had intended to go back and handle that case and then forgot about it.

@huangp Another small change to review, please.

**Comment by huangp** [comment]
:lgtm:

---

Reviewed 14 of 14 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/18#-:-KWBfjWqcYbg_PdP2IXn:bnfp4nl)*

**Comment by djansen-redhat** [comment]
:+1: good job!

**Pull request merged**